### DISCUSSION TITLE
Cannot push to main but can push to master

### DISCUSSION THREAD

**Discussion by OnlineTeachingNow** [community/community#22512] [New to GitHub]
<p>Hello, all. I am new to Github and have been carefully following several tutorials in order to create my first repository that results in nothing but failure and confusion.</p>
<p>When I create a repository, I see that the only branch is labeled as “main”. However, in Git bash, I never see main, my directory is instead labeled “(master)”. However, I assume this is some hangover from this new PC change, so I decided to do everything by referring to the ‘main’ branch, which I have read is no different from the master and is the societally approved way of referring to the most important branch going forward.</p>
<p>I do all my pull/commit commands with main, i.e. git pull origin main and it seems to work. However, in the stage of pushing (git push origin main), I get this error: error: src refspec main does not match any<br>
error: failed to push some refs to ‘<a href="https://github.com/blah" class="inline-onebox" rel="noopener nofollow ugc">BLAH · GitHub</a> blah blah.git’</p>
<p>I have read over and over again that this is because I didn’t commit properly. However, I commit again and it says: On branch master nothing to commit, working tree clean (so, again, it refers to a master branch that doesn’t exist).</p>
<p>So, I changed my push to “git push origin master” and now it adds a master branch to my repository and adds all the project files as expected to that branch. But, now my repository is clearly racist, which implicates me and my products as racist as well.</p>
<p>My question is what in the heck is going on? What is the difference between main and master? Why can I not commit to main but can create a master branch and commit to that? Why does Git Bash refer to my main branch as a master branch? If I create a master branch, is the main branch still the master one?</p>

**Comment by jfullstackdev** [comment]
<p>you may further install GitHub Desktop, through GitHub Desktop, your problem might be solved, it’s just like it’s the middle man between <code>git</code> ang <code>GitHub</code>.</p>
<p>Now, regarding your question, there is no difference. But GitHub decided that instead of <code>master</code>, it should be called <code>main</code> because the term <code>master</code> indicates slavery. So, they decided to change that to <code>main</code>.</p>

**Reply by zapbra** [reply]
what a joke

**Comment by OnlineTeachingNow** [comment]
<p>Thanks for your response. However, there is clearly a difference between Main and Master because I can only use main sometimes and not other times. For example:</p>
<ol>
<li>
<p>I create a new repository and try to pull from the master<br>
$ git pull origin master<br>
fatal: couldn’t find remote ref master</p>
</li>
<li>
<p>So, then I try to pull from main, which is successful.<br>
$ git pull origin main<br>
remote: Enumerating objects: 3, done.<br>
remote: Counting objects: 100% (3/3), done.<br>
remote: Compressing objects: 100% (2/2), done.<br>
remote: Total 3 (delta 0), reused 0 (delta 0), pack-reused 0<br>
Unpacking objects: 100% (3/3), 1.09 KiB | 23.00 KiB/s, done.<br>
From <a href="https://github.com/OnlineTeachingNow/TileVania" class="inline-onebox" rel="noopener nofollow ugc">GitHub - OnlineTeachingNow/TileVania: 2D Tile game created in Gamedev.tv's Unity course</a></p>
</li>
</ol>
<ul>
<li>branch            main       → FETCH_HEAD</li>
<li>[new branch]      main       → origin/main</li>
</ul>
<ol start="3">
<li>However, when I check the available branches, it says there is only a master branch!<br>
$ git branch</li>
</ol>
<ul>
<li>master</li>
</ul>
<ol start="4">
<li>
<p>I commit the files.</p>
</li>
<li>
<p>I push to the main, which gives me an error.<br>
$ git push origin main<br>
error: src refspec main does not match any<br>
error: failed to push some refs to ‘<a href="https://github.com/OnlineTeachingNow/TileVania.git" rel="noopener nofollow ugc">https://github.com/OnlineTeachingNow/TileVania.git</a>’</p>
</li>
<li>
<p>So, I push to the master, which is successful.<br>
$ git push origin master<br>
Enumerating objects: 5031, done.<br>
Counting objects: 100% (5031/5031), done.<br>
Delta compression using up to 8 threads<br>
Compressing objects: 100% (5017/5017), done.<br>
Writing objects: 100% (5030/5030), 17.01 MiB | 7.67 MiB/s, done.<br>
Total 5030 (delta 2730), reused 0 (delta 0), pack-reused 0<br>
remote: Resolving deltas: 100% (2730/2730), done.<br>
remote:<br>
remote: Create a pull request for ‘master’ on GitHub by visiting:<br>
remote:      url<br>
remote:<br>
url</p>
</li>
</ol>
<ul>
<li>[new branch]      master → master</li>
</ul>
<p>This is absolutely ridiculous! What is the difference between these two? Why can I only pull from main but push to master?</p>

**Reply by noahostle** [reply]
4 years later and this behavior still occurs, why dont they change the default local branch name to main! so annoying

**Comment by airtower-luna** [comment, marked as answer]
<p>The difference is the name: Your local branch is called <code>master</code>, the remote one <code>main</code>. By default <code>git push origin &lt;branch&gt;</code> will push that branch to a remote branch <em>of the same name</em>. If you really want to you can override that and explicitly push from your local <code>master</code> to remote <code>main</code>:</p>
<pre><code class="lang-bash">git push origin master:main
</code></pre>
<p>However, it might be much easier to rename your local branch to match the remote one:</p>
<pre><code class="lang-bash">git branch -m master main
</code></pre>

**Reply by Gaalpos** [reply]
thanks for the help!!

**Reply by beelalamin** [reply]
Before Push rebase then push : 

`git pull --rebase origin main`

**Reply by gasaiaaa4** [reply]
> The difference is the name: Your local branch is called `master`, the remote one `main`. By default `git push origin <branch>` will push that branch to a remote branch _of the same name_. If you really want to you can override that and explicitly push from your local `master` to remote `main`:
> 
> ```
> git push origin master:main
> ```
> 
> However, it might be much easier to rename your local branch to match the remote one:
> 
> ```
> git branch -m master main
> ```

This is what i was looking for

**Comment by OnlineTeachingNow** [comment]
<p>OK, thank you for the clarification. Would I have to rename the local master branch every time I create a new project? Why is the default on my local branch called ‘master’? I did not create that. Is that the default of Git Bash? Aren’t Git Hub and Bash made by the same people? Why did they decide to be confusingly “antiracist” in one thing, but maintain the “racist” status quo in the other program? The least they could have done is implement consistency.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="OnlineTeachingNow" data-post="5" data-topic="161967">
<div class="title">
<div class="quote-controls"></div>
 OnlineTeachingNow:</div>
<blockquote>
<p>Aren’t Git Hub and Bash made by the same people?</p>
</blockquote>
</aside>
<p>Nope. GitHub is a company building on Git. I’m pretty sure some GitHub employees are contributing to Git, but it’s separate.</p>
<p>Git somewhat recently (version 2.28) introduced the <a href="https://git-scm.com/docs/git-config#Documentation/git-config.txt-initdefaultBranch"><code>init.defaultBranch</code> option</a> to let you configure the name of the default branch on init, with the eventual goal of moving away from <code>master</code> (after making sure nothing relies on that name).</p>

**Comment by OnlineTeachingNow** [comment]
<p>Got it. Thanks so much for answering my questions. I’ll be renaming the local branch for future projects.</p>

**Comment by Momo979** [comment]
<p>Thank you, this was really annoying me!</p>

**Comment by wzm2256** [comment]
Thanks for your answer. I have been struggling to fight this error for one hour. My local default branch is main, but the remote overleaf branch is still master, so strange errors keep showing up! I will rename my local branch to solve this.

Why are all the fuss about master and main? Github should do something more meaningful than just changing those boring names if they really want to stop racisim or other stuffs!

**Comment by Kaushan-Dutta** [comment]
If the above solution still not works then make a force push. And since this problem will occur just during the first time push no loss of data will occur.
